cpp/include/cudf/groupby.hpp

```diff
@@ -0,0 +1,223 @@
+      results{};
+};
+
+/**
+ * @brief Groups values by keys and computes aggregations on those groups.
+ */
+class groupby {
+ public:
+  groupby() = delete;
+  ~groupby() = default;
+  groupby(groupby const&) = delete;
+  groupby(groupby&&) = delete;
+  groupby& operator=(groupby const&) = delete;
+  groupby& operator=(groupby&&) = delete;
+
+  /**
+   * @brief Construct a groupby object with the specified `keys`
+   *
+   * @note This object does *not* maintain the lifetime of `keys`. It is the
+   * users responsibility to ensure the `groupby` object does not outlive
+   * `keys`.
+   *
+   * @param keys Table whose rows act as the groupby keys
+   * @param ignore_null_keys Indicates whether rows in `keys` that contain NULL
+   * values should be ignored
+   * @param keys_are_sorted Indicates whether rows in `keys` are already sorted
+   * @param column_order If `keys_are_sorted == true`, indicates whether each
+   * column is ascending/descending. If empty, assumes all  columns are
+   * ascending. Ignored if `keys_are_sorted == false`.
+   * @param null_precedence If `keys_are_sorted == true`, indicates the ordering
+   * of null values in each column. Else, ignored. If empty, assumes all columns
+   * use `null_order::BEFORE`. Ignored if `keys_are_sorted == false`.
+   */
+  explicit groupby(table_view const& keys, bool ignore_null_keys = true,
+                   bool keys_are_sorted = false,
+                   std::vector<order> const& column_order = {},
+                   std::vector<null_order> const& null_precedence = {});
+
+  /**
+   * @brief Performs grouped aggregations on the specified values.
+   *
+   * The values to aggregate and the aggregations to perform are specifed in an
+   * `aggregation_request`. Each request contains a `column_view` of values to
+   * aggregate and a set of `aggregation`s to perform on those elements.
+   *
+   * For each `aggregation` in a request, `values[i]` will be aggregated with
+   * all other `values[j]` where rows `i` and `j` in `keys` are equivalent.
+   *
+   * The `size()` of the request column must equal `keys.num_rows()`.
+   *
+   * @throws cudf::logic_error If `requests[i].values.size() !=
+   * keys.num_rows()`.
+   *
+   * Example:
+   * ```
+   * Input:
+   * keys:     {1 2 1 3 1}
+   *           {1 2 1 4 1}
+   * request:
+   *   values: {3 1 4 9 2}
+   *   aggregations: {{SUM}, {MIN}}
+   *
+   * result:
+   *
+   * keys:  {3 1 2}
+   *        {4 1 2}
+   * values:
+   *   SUM: {9 9 1}
+   *   MIN: {9 2 1}
+   * ```
+   *
+   * @param requests The set of columns to aggregate and the aggregations to
+   * perform
+   * @param mr Memory resource used to allocate the returned table and columns
+   * @return Pair containing the table with each groups unique key and
+   * a set of aggregation_results for each request in the same order as was
+   * specified in `requests`.
+   */
+  std::pair<std::unique_ptr<table>, std::vector<aggregation_result>> aggregate(
```

Comment:
Ideally I'd like it if this just returned `std::vector<aggregation_result>` and the corresponding keys could be accessed via `groupby::unique_keys()`, however, I'm concerned about ensuring that the ordering of the values in the aggregation results and the unique keys are consistent if they are split into two functions.

Reply:
No, but the resulting order of keys is indeterminate if a hash-based groupby is used. 

My concern is that the result of `unique_keys()` could change from one invocation to the next. For example:
```
groupby g(...);

results = g.aggregate( /* hash aggregations */); // uses hash based groupby
unique_keys = g.unique_keys(); // the order of unique_keys should match the order of results (random order)

new_result = g.aggregate( /* quantile */); // uses hash groupby
new_unique_keys = g.unique_keys(); // the order of new_unique_keys should match the order of new_result (sorted order)
// However, the order of unique_keys != the order of new_unique_keys. This seems like a problem
```

Reply:
One idea discussed with @harrism was instead of returning a table of the unique keys from `groupby::aggregate` we could return a column of unique group IDs. This column of IDs could be used to materialize the actual unique keys that correspond to each group.

So for example:
```
[IDs, aggregation_results] = g.aggregate(...);

group_keys = g.unique_keys(IDs);
```

However, I don't really love this idea either. 

The other alternative is to require that `groupby::aggregate` always return the values in order that corresponds to the sorted order of the unique keys. This would be a unnecessary performance overhead for a hash-based implementation.

Reply:
Option 2: (and this is made possible by the groupby object)
Don't allow the aforementioned aggregations in the `aggregate()` method. This can be done by not making an args struct for them. Then they'd need to be done using a separate `groupby().value_counts()` method call. This method can take all the args needed and can return a key table that better suits the needs of this kind of aggregation. Obvious downside is that it may need a re-sort of the value column if that column was included in a previous `aggregate()` call.

Reply:
> Take, for example, unique, nlargest (which returns less than n values if there aren't enough, for a key) and value_counts (histogram).

Ideally for all of these the returned values would be a column of lists, such that there is still only one "thing" returned per key (in this case the "thing" just happens to be a list).